### PR TITLE
`<format>`: Copy `basic_format_args` in `tuple` formatters when needed

### LINKED ISSUE #4651: `<format>`: Underlying formatters of `pair-or-tuple` formatter cannot access format args
Underlying formatters of [`pair-or-tuple`](http://eel.is/c++draft/format.tuple) formatter should be able to call `format_ctx.arg(I)` and get correct format arg.

Example:

```c++
#include <print>
#include <thread>

template <size_t I> struct substitute_arg {};

template <size_t I> struct std::formatter<substitute_arg<I>> {
  template <class ParseContext> constexpr auto parse(ParseContext &ctx) {
    auto it = ctx.begin();
    if (it != ctx.end() && *it != '}') {
      throw format_error{"Expected empty spec"};
    }

    ctx.check_arg_id(I);
    return it;
  }

  template <class FormatContext>
  auto format(substitute_arg<I>, FormatContext &ctx) const {
    auto visitor = [&]<class T>(T val) -> FormatContext::iterator {
      if constexpr (same_as<T, monostate>) {
        return ranges::copy("monostate"sv, ctx.out()).out;
      } else if constexpr (same_as<T, typename basic_format_arg<
                                          FormatContext>::handle>) {
        format_parse_context parse_ctx{""};
        val.format(parse_ctx, ctx);
        return ctx.out();
      } else {
        return format_to(ctx.out(), "{}", val);
      }
    };

    return visit_format_arg(visitor, ctx.arg(I));
  }
};

int main() {
  std::println("{0:}", std::tuple{substitute_arg<1>{}, substitute_arg<2>{}},
               "thread::id", std::thread::id{});
}
```

Expected output (libc++: https://godbolt.org/z/1c4fdhzTb):

```console
(thread::id, 0)
```

We've got:

```console
(monostate, monostate)
```

### PR DESCRIPTION
**Pull request by frederick-vs-ja**
Fixes #4651.

Also
- avoids creating a temporary string when the minimum width is specified to 0 or not specified, completing the "Future work" in #4631;
- reorganizes `_Fmt_it`, `_Fmt_wit`, `format_context`, and `wformat_context` with consistent alias templates; and
- tests that functions taking non-constructible `basic_format_context` specializations can still be well-formed, despite that they can't be actually called.

Notes: LWG-4061 is now tentatively ready (although this PR doesn't speculatively implements it yet). Thus it should be now clear that `basic_format_context` objects can only be created by implementations, not user codes.

### CODE REVIEW

**Comment by cpplearner** [comment]
This doesn't really copy the arguments, right? It copies `basic_format_args`, which holds a pointer to the array of `format_arg` objects.

File: stl/inc/format [hunk 1]
```diff
@@ -4121,26 +4142,26 @@ protected:
...
-                }
-                _STD get<_Idx>(_Underlying).format(_Arg, _Tmp_ctx);
-            };
-            (_Single_writer.template operator()<_Indices>(_Args), ...);
-        }(index_sequence_for<_ArgTypes...>{});
-        _STD _Copy_unchecked(_Closing_bracket._Unchecked_begin(), _Closing_bracket._Unchecked_end(), _Tmp_ctx.out());
+        basic_string<_CharT> _Tmp_str;
+        if constexpr (is_same_v<_FormatContext, _Default_format_context<_CharT>>) {
+            _Fmt_iterator_buffer<back_insert_iterator<basic_string<_CharT>>, _CharT> _Tmp_buf{
+                _STD back_inserter(_Tmp_str)};
```
Comment (StephanTLavavej) [inline stl/inc/format:4153]: Pre-existing style nitpick: The rest of `<format>` uses `back_insert_iterator` CTAD, which skips a function call in debug mode.

File: tests/std/tests/P2286R8_text_formatting_tuple/test.cpp [hunk 2]
```diff
@@ -435,10 +435,32 @@ auto test_vformat_exception = []<class CharT, class... Args>([[maybe_unused]] st
     }
 };
 
+// Also test that functions taking non-constructible basic_format_context specializations can be well-formed,
+// despite that they can't be actually called.
```
Comment (StephanTLavavej) [inline tests/std/tests/P2286R8_text_formatting_tuple/test.cpp:439]: Grammar nitpicks: "despite that" doesn't sound right; I would expect "despite the fact that". (https://en.wiktionary.org/wiki/despite notes this.) Also, "can't be actually called" should be "can't actually be called". English is super weird :joy_cat:

File: tests/std/tests/P2286R8_text_formatting_tuple/test.cpp [hunk 3]
```diff
@@ -435,10 +435,32 @@ auto test_vformat_exception = []<class CharT, class... Args>([[maybe_unused]] st
...
+
+template <class CharT>
+void test_unconstructible_format_context_for_raw_ptr(basic_format_context<CharT*, CharT>& ctx) { // COMPILE-ONLY
+    formatter<tuple<basic_string<CharT>>, CharT> formatter;
+    formatter.format(make_tuple(basic_string<CharT>(STR("42"))), ctx);
+}
+
+template <class CharT>
+void test_unconstructible_format_context_for_back_inserter(
+    basic_format_context<back_insert_iterator<basic_string<CharT>>, CharT>& ctx) { // COMPILE-ONLY
```
Comment (StephanTLavavej) [inline tests/std/tests/P2286R8_text_formatting_tuple/test.cpp:449]: Should include `<iterator>`.

File: tests/std/tests/P2286R8_text_formatting_tuple/test.cpp [hunk 4]
```diff
@@ -435,10 +435,32 @@ auto test_vformat_exception = []<class CharT, class... Args>([[maybe_unused]] st
     }
 };
 
+// Also test that functions taking non-constructible basic_format_context specializations can be well-formed,
+// despite that they can't be actually called.
+
+template <class CharT>
+void test_unconstructible_format_context_for_raw_ptr(basic_format_context<CharT*, CharT>& ctx) { // COMPILE-ONLY
+    formatter<tuple<basic_string<CharT>>, CharT> formatter;
```
Comment (StephanTLavavej) [inline tests/std/tests/P2286R8_text_formatting_tuple/test.cpp:443]: `formatter` is shadowing.

File: tests/std/tests/P2693R1_text_formatting_thread_id/test.cpp [hunk 5]
```diff
@@ -254,12 +256,62 @@ void check_invalid_specs() {
...
+        return visit_format_arg(visitor, ctx.arg(I));
+    }
+};
+
+template <class CharT>
+void check_substitute_arg_with_tuple_formatters() {
+    assert(format(STR("{0:}"), tuple{substitute_arg<1>{}, substitute_arg<2>{}}, STR("thread::id"), thread::id{})
+           == STR("(thread::id, 0)"));
+    assert(format(STR("{0:}"), pair{substitute_arg<1>{}, substitute_arg<2>{}}, STR("thread::id"), thread::id{})
+           == STR("(thread::id, 0)"));
```
Comment (StephanTLavavej) [inline tests/std/tests/P2693R1_text_formatting_thread_id/test.cpp:304]: Should include `<tuple>` and `<utility>`.

File: tests/std/tests/P2693R1_text_formatting_thread_id/test.cpp [hunk 6]
```diff
@@ -254,12 +256,62 @@ void check_invalid_specs() {
...
+
+        ctx.check_arg_id(I);
+        return it;
+    }
+
+    template <class FormatContext>
+    auto format(substitute_arg<I>, FormatContext& ctx) const {
+        auto visitor = [&]<class T>(T val) -> FormatContext::iterator {
+            if constexpr (same_as<T, monostate>) {
+                return ranges::copy(STR("monostate"sv), ctx.out()).out;
```
Comment (StephanTLavavej) [inline tests/std/tests/P2693R1_text_formatting_thread_id/test.cpp:285]: Should include `<string_view>`.

**Comment by StephanTLavavej** [comment]
Thanks! I pushed a conflict-free merge with `main` and some minor nitpicks.

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
Thanks for fixing the runtime correctness of one of the most popular new features! :magic_wand: :cat2: :cat:

**Pull request merged**